src/core/macro.scala

```diff
@@ -259,12 +281,223 @@ object Macro:
         .groupBy { case (name, _) => name }
         .toList
         .map { case (name, l) => name -> l.flatMap(_._2) }
-        .map { (name, anns) => Expr.ofTuple(Expr(name), Expr.ofList(anns)) }
+        .map { case (name, anns) => Expr.ofTuple(Expr(name), Expr.ofList(anns)) }
 
     private def isObjectOrScala(bc: Symbol) =
       bc.name.contains("java.lang.Object") || bc.fullName.startsWith("scala.")
 
     private def filterAnnotation(a: Term): Boolean =
       a.tpe.typeSymbol.maybeOwner.isNoSymbol ||
         a.tpe.typeSymbol.owner.fullName != "scala.annotation.internal"
-  }
+
+object MirrorlessMacro:
+  
+  object ValueClassDerivation:
+
+    inline def getValueClassParam[TC[_], V]: CaseClass.Param[TC, V] = 
+     ${ getValueClassParamImpl[TC, V] }
+
+    inline def rawContructByMacro[V](as: List[Any]): V = 
+      ${ rawContructByMacroImpl[V]('as) } 
+
+    def getValueClassParamImpl[TC[_]: Type, A: Type](using Quotes): Expr[CaseClass.Param[TC, A]] = 
+      import quotes.reflect.* 
+      
+      // repeated
+      def isObjectOrScala(bc: Symbol) =
+        bc.name.contains("java.lang.Object") || bc.fullName.startsWith("scala.")
+
+      def fromConstructor(from: Symbol): List[(String, List[Term])] = 
+        from.primaryConstructor.paramSymss.flatten.map { field =>
+          field.name -> field.annotations
+            .filter(filterAnnotation)
+        }
+
+      def fromDeclarations(from: Symbol): List[(String, List[Term])] = 
+        from.declarations.collect {
+          case field: Symbol if field.isValDef => 
+            field.name -> field.annotations.filter(filterAnnotation)
+        }
+
+      def groupTermsByName(tss: List[(String, List[Term])]): List[(String, List[Term])] = 
+        tss
+          .groupBy { case (name, _) => name }
+          .toList
+          .map { case (name, l) => name -> l.flatMap(_._2) }
+
+      def selectTermsByName(tss: List[(String, List[Term])], name: String): List[Term] = 
+        tss
+          .toMap
+          .getOrElse(name, Nil)
+      
+      def filterAnnotation(annTerm: Term): Boolean =
```

Comment:
this looks the same as a private method above, cannot we share it?

Reply:
Indeed - it was due to  Quotes being path-dependent and I had some troubles with accessing that. Now fixed and shared .